### ISSUE TITLE
Feature Request: Add support for ScatterElements operator during compilation

### ISSUE THREAD

**Issue by fabecode** [zama-ai/concrete-ml#531]
## Summary
I have implemented a GNN model, quantised using Brevitas, and compiled using 'compile_brevitas_qat_model'. 
During the compilation, I face the following error: "ValueError: The following ONNX operators are required to convert the torch model to numpy but are not currently implemented: ScatterElements."

Would it be possible to add ScatterElements as a supported operator, or is there any suggested workarounds? Thank you!

**Comment by bcm-at-zama** [comment]
Hello, the team is remote for the end of the week, so we'll answer next week. Would it be possible you share your code, to simplify the reproduction of the error? Or, at least, could you show us your model with Brevitas layers?

**Comment by bcm-at-zama** [comment]
Great, we'll have a look! As I said, it's for next week, you'll have to be a bit patient :)

**Comment by andrei-stoian-zama** [comment]
Hi! thanks for sending the code! Great job converting your network to Brevitas. Here are a few answers:

- While the code gives us good information, having a reproducible code that shows the inputs (x, edge_index,  that you pass to your model would help. It would allow us to inspect the ONNX that is generated. Can you add that information please?
- We have not yet implemented ScatterElements, which I guess comes from `x = x[edge_index.T].reshape(-1, 2 * self.n_hidden).relu()`. It would take some time to do and it would only work if you index all of the elements in `x` with that call. Is that the case?
- You also concatenate two tensors here: `        x = torch.cat((x, edge_attr.view(-1, edge_attr.shape[1])), 1)` . This will probably fail with quantization since the two concatenated tensors will not have matching quantization parameters. This can be fixed by instantiating a QuantIdentity layer in your model init  and applying it to both inputs that are concatenated. 




`

**Comment by bcm-at-zama** [comment]
And obviously @fabecode, for the information @andrei-stoian-zama is asking you: you can send to my email address if you prefer things to be confidential. Just make sure it opens well, we've had issues before.

**Comment by fabecode** [comment]
Hi Andrei and Benoit, thank you for the prompt reply! Here are my responses to the questions:
- I have emailed the link to my code base as well as the ONNX file of my quantised model to @bcm-at-zama .
- I believe so. Before the reshape, here are my input shapes - `x: torch.Size([9142, 66])` and `edge_attr: torch.Size([5851, 66])`
- Thank you for the advice! Could you kindly advise where to insert the QuantIdentity layer? When I tried to add QuantIdentity, I faced the "The quantizer passed does not adhere to the quantization protocol" error.

**Comment by fabecode** [comment]
Hello @andrei-stoian-zama @bcm-at-zama, I would also like to inquire if there is an anticipated timeline for the implementation of Scatter Elements? And if there is any way I could try to contribute to the implementation process? Thank you!

**Comment by andrei-stoian-zama** [comment]
Hello again and thank you for the code ! 

Unfortunately I can't give a deadline for this. It would take a bit of work in both Concrete ML and Concrete.

**Comment by fabecode** [comment]
I see, thanks again!

**Comment by bcm-at-zama** [comment]
Yes @fabecode , sorry but it's certainly a complicated task to do for an external person. We should have it in 2024, hopefully! 

If you want to contribute, can we encourage you to participate to our bounties? https://github.com/zama-ai/bounty-and-grant-program

**Comment by fabecode** [comment]
Thank you, will be happy to participate in the bounties when I have the time! 

Actually, the GNN project I have embarked on is part of my undergraduate final thesis which is due very soon, and my university has just given the green light for me to apply for the Zama Grant Program. The final leg would be the compilation of the model to be TFHE compatible, so would be really grateful if you happen to have suggestions on temporary workarounds to the ScatterElements error that I can try out 🙏

**Comment by fabecode** [comment]
Yes, I just checked, `edge_index` is a constant!

**Comment by fabecode** [comment]
Thank you so much @andrei-stoian-zama for the suggestions, will try it out!

**Comment by fabecode** [comment]
After further analysis, below are my findings:
- `edge_index` does not have the same length as `x`.
- In `x = x[edge_index.T].reshape(-1, 2 * self.n_hidden).relu()` line:
  - Tensor shape of `edge_index`: [2, num_edges]
    - where 2 represents source + destination nodes
  - Tensor shape of `x` before reshape: [num_nodes, num_features]
  - So, tensor shape of `x[edge_index.T]`: [num_edges, 2, num_features]
    - where `x[edge_index.T]` is a tensor where each row contains the features of the source and dest nodes in each edge
  - Tensor shape of `x[edge_index.T].reshape(-1, 2*self.n_hidden)`: [num_edges, 2*self.n_hidden]
  - Tensor shape of `x` after relu: [num_edges, 2*self.n_hidden]
- In `x = torch.cat((x, edge_attr.view(-1, edge_attr.shape[1])), 1)` line:
  - Tensor shape of `edge_attribute`: [num_edges, num_features]
  - Tensor shape of `(edge_attr.view(-1, edge_attr.shape[1]))`: [num_edges, num_features]
  - Tensor shape of `(torch.cat((x, edge_attr.view(-1, edge_attr.shape[1])), 1))`: [num_edges, num_features + 2*self.n_hidden]

Also, upon looking at the ONNX graph, it might be possible that ScatterElements occur in the `torch.cat` line instead of the `reshape` line? (You could view the full graph in the ONNX file I sent in the earlier email.)
<img width="464" alt="image" src="https://github.com/zama-ai/concrete-ml/assets/65911192/5cd1a0ee-d255-4178-87ce-045d619018da">

Could you kindly advise? Thank you once again!

**Comment by andrei-stoian-zama** [comment]
I'm afraid I don't know how to help more on this. 

As a first step you should try to determine which torch operator produces ScatterElements. I don't think it's the concatenation though.

**Comment by fabecode** [comment]
No worries, I really appreciate all your advice thus far! I will do my best to find a temporary alternative to ScatterElements.

**Comment by tobiasroth04** [comment]
Hello,

I am working on the above-mentioned project and have encountered the same error. Is there a known solution to this problem yet?

I would greatly appreciate a response.
